packages/client/src/perma-cache.js

```diff
@@ -0,0 +1,396 @@
+ * @typedef { import('./lib/interface.js').CacheResult} CacheResult
+ * @typedef { import('./lib/interface.js').CacheDeleteResult} CacheDeleteResult
+ * @typedef { import('./lib/interface.js').CacheEntry} CacheEntry
+ * @typedef { import('./lib/interface.js').AccountInfo} AccountInfo
+ */
+
+/**
+ * @implements Service
+ */
+export class PermaCache {
+  /**
+   * Constructs a client bound to the given `options.token` and
+   * `options.endpoint`.
+   *
+   * @example
+   * ```js
+   * import { PermaCache } from 'nftstorage.link'
+   * const client = new PermaCache({ token: API_TOKEN })
+   * ```
+   *
+   * @param {{token: string, endpoint?:URL, rateLimiter?: RateLimiter}} options
+   */
+  constructor({
+    token,
+    endpoint = new URL('https://api.nftstorage.link'),
+    rateLimiter,
+  }) {
+    /**
+     * Authorization token.
+     *
+     * @readonly
+     */
+    this.token = token
+    /**
+     * Service API endpoint `URL`.
+     * @readonly
+     */
+    this.endpoint = endpoint
+    /**
+     * @readonly
+     */
+    this.rateLimiter = rateLimiter || createRateLimiter()
+  }
+
+  /**
+   * @hidden
+   * @param {string} token
+   * @returns {Record<string, string>}
+   */
+  static headers(token) {
+    /* c8 ignore next 1 */
+    if (!token) throw new Error('missing token')
+    return {
+      Authorization: `Bearer ${token}`,
+      'X-Client': 'nftstorage.link/js',
+    }
+  }
+
+  /**
+   * @param {Service} service
+   * @param {string[]} urls
+   * @param {PutOptions} [options]
+   * @returns {Promise<CacheResult[]>}
+   */
+  static async put(
+    { endpoint, token, rateLimiter = globalRateLimiter },
+    urls,
+    { onPut, maxRetries = MAX_RETRIES } = {}
+  ) {
+    areNftStorageLinkUrls(urls)
+
+    const headers = PermaCache.headers(token)
+
+    /** @type {import('p-settle').PromiseResult<CacheResult>[]} */
+    const cacheResults = await pSettle(
+      urls.map(async (url) => {
+        const apiUrl = new URL(
+          `perma-cache/${encodeURIComponent(url)}`,
+          endpoint
+        )
+
+        let cacheResult
+        try {
+          cacheResult = await pRetry(
+            async () => {
+              await rateLimiter()
+              const request = await fetch(apiUrl.toString(), {
+                method: 'POST',
+                headers,
+              })
+
+              /** @type {CacheResult} */
+              const res = await request.json()
+              if (!request.ok) {
+                // @ts-ignore Only exists on Error
+                const e = new Error(res.message)
+                // do not retry if abortable errors - will not succeed
+                if (ABORTABLE_ERRORS_STATUS.includes(request.status)) {
+                  throw new AbortError(e)
+                }
+                /* c8 ignore next 2 */
+                throw e
+              }
+
+              onPut && onPut(url)
+
+              return {
+                ...res,
+                url, // decoded
+              }
```

Comment:
`url` should be in the response no?

```suggestion
              return res
```

Reply:
yeah, the issue was in the mock 🤦🏼